File: Linq/Extensions.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using Raven.Json.Utilities;

namespace Raven.Json.Linq
{
	public static class Extensions
	{
		/// <summary>
		/// Converts the value.
		/// </summary>
		/// <typeparam name="U">The type to convert the value to.</typeparam>
		/// <param name="value">A <see cref="RavenJToken"/> cast as a <see cref="IEnumerable{T}"/> of <see cref="RavenJToken"/>.</param>
		/// <returns>A converted value.</returns>
		public static U Value<U>(this IEnumerable<RavenJToken> value)
		{
			return value.Value<RavenJToken, U>();
		}

		public static U Value<U>(this RavenJToken value)
		{
			return value.Convert<U>();
		}

		/// <summary>
		/// Converts the value.
		/// </summary>
		/// <typeparam name="T">The source collection type.</typeparam>
		/// <typeparam name="U">The type to convert the value to.</typeparam>
		/// <param name="value">A <see cref="RavenJToken"/> cast as a <see cref="IEnumerable{T}"/> of <see cref="RavenJToken"/>.</param>
		/// <returns>A converted value.</returns>
		public static U Value<T, U>(this IEnumerable<T> value) where T : RavenJToken
		{
			ValidationUtils.ArgumentNotNull(value, "source");

			var token = value as RavenJToken;
			if (token == null)
				throw new ArgumentException("Source value must be a RavenJToken.");

			return token.Convert<U>();
		}

		/// <summary>
		/// Returns a collection of converted child values of every object in the source collection.
		/// </summary>
		/// <typeparam name="U">The type to convert the values to.</typeparam>
		/// <param name="source">An <see cref="IEnumerable{T}"/> of <see cref="RavenJToken"/> that contains the source collection.</param>
		/// <returns>An <see cref="IEnumerable{T}"/> that contains the converted values of every node in the source collection.</returns>
		public static IEnumerable<U> Values<U>(this IEnumerable<RavenJToken> source)
		{
			return Values<U>(source, null);
		}

		/// <summary>
		/// Returns a collection of child values of every object in the source collection with the given key.
		/// </summary>
		/// <param name="source">An <see cref="IEnumerable{T}"/> of <see cref="RavenJToken"/> that contains the source collection.</param>
		/// <param name="key">The token key.</param>
		/// <returns>An <see cref="IEnumerable{T}"/> of <see cref="RavenJToken"/> that contains the values of every node in the source collection with the given key.</returns>
		public static IEnumerable<RavenJToken> Values(this IEnumerable<RavenJToken> source, string key)
		{
			return Values<RavenJToken>(source, key);
		}

		/// <summary>
		/// Returns a collection of child values of every object in the source collection.
		/// </summary>
		/// <param name="source">An <see cref="IEnumerable{T}"/> of <see cref="RavenJToken"/> that contains the source collection.</param>
		/// <returns>An <see cref="IEnumerable{T}"/> of <see cref="RavenJToken"/> that contains the values of every node in the source collection.</returns>
		public static IEnumerable<RavenJToken> Values(this IEnumerable<RavenJToken> source)
		{
			return source.Values(null);
		}

		internal static IEnumerable<U> Values<U>(this IEnumerable<RavenJToken> source, string key)
		{
			ValidationUtils.ArgumentNotNull(source, "source");

			foreach (RavenJToken token in source)
			{
				if (token is RavenJValue)
				{
					yield return Convert<U>(token);
				}
				else
				{
					foreach (RavenJToken t in token.Children())
					{
						yield return t.Convert<U>();
					}
				}

				var ravenJObject = (RavenJObject) token;

				RavenJToken value = ravenJObject[key];
				if (value != null)
					yield return value.Convert<U>();
			}

			yield break;
		}

		internal static U Convert<U>(this RavenJToken token)
		{
			if (token is RavenJArray && typeof(U) == typeof(RavenJObject))
			{
				var ar = (RavenJArray)token;
				var o = new RavenJObject();
				foreach (RavenJObject item in ar.Items)
				{
					o[item["Key"].Value<string>()] = item["Value"];
				}
				return (U) (object) o;
			}

			bool cast = typeof(RavenJToken).IsAssignableFrom(typeof(U));

			return Convert<U>(token, cast);
		}

		internal static IEnumerable<U> Convert<U>(this IEnumerable<RavenJToken> source)
		{
			ValidationUtils.ArgumentNotNull(source, "source");

			bool cast = typeof(RavenJToken).IsAssignableFrom(typeof(U));

			foreach (RavenJToken token in source)
			{
				yield return Convert<U>(token, cast);
			}
		}

		internal static U Convert<U>(this RavenJToken token, bool cast)
		{
			if (cast)
			{
				// HACK
				return (U)(object)token;
			}
			else
			{
				if (token == null)
					return default(U);

				var value = token as RavenJValue;
				if (value == null)
					throw new InvalidCastException("Cannot cast {0} to {1}.".FormatWith(CultureInfo.InvariantCulture, token.GetType(), typeof(U)));

				if (value.Value is U)
					return (U)value.Value;

				Type targetType = typeof(U);

				if (targetType.IsGenericType && targetType.GetGenericTypeDefinition() == typeof(Nullable<>))
				{
					if (value.Value == null)
						return default(U);

					targetType = Nullable.GetUnderlyingType(targetType);
				}

				return (U)System.Convert.ChangeType(value.Value, targetType, CultureInfo.InvariantCulture);
			}
		}
	}
}

File: Raven.Client.Java/lib/src/main/java/raven/abstractions/json/linq/Extensions.java
package raven.abstractions.json.linq;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.UUID;

import org.apache.commons.beanutils.ConvertUtils;

import raven.abstractions.data.Constants;

public class Extensions {

  public static <U> U value(Class<U> clazz, Iterable<RavenJToken> value) {
    return value(RavenJToken.class, clazz, value);
  }

  public static <U> U value(Class<U> clazz, RavenJToken value) {
    return convert(clazz, value);
  }

  public static <U> Iterable<U> values(Class<U> clazz, Iterable<RavenJToken> source) {
    return values(clazz, source, null);
  }

  public static Iterable<RavenJToken> values(Iterable<RavenJToken> source, String key) {
    return values(RavenJToken.class, source, key);
  }

  public static Iterable<RavenJToken> values(Iterable<RavenJToken> source) {
    return values(source, null);
  }

  //TODO: fixme
  private static <U> Iterable<U> values(Class<U> clazz, Iterable<RavenJToken> source, String key) {
    return new RavenJTokenIterable(clazz, source, key);
  }

  public static <T extends RavenJToken, U> U value(Class<T> clazzT, Class<U> clazzU, Iterable<T> value) //where T : RavenJToken
  {
    RavenJToken token = null;
    try {
      token = (RavenJToken) value;
    } catch (ClassCastException e) {
      //
    }
    if (token == null) throw new IllegalArgumentException("Source value must be a RavenJToken.");

    return convert(clazzU, token);
  }

  private static <U> U convert(Class<U> clazz, RavenJToken token) {
    if (token instanceof RavenJArray && RavenJObject.class.equals(clazz)) {
      RavenJArray ar = (RavenJArray) token;
      RavenJObject o = new RavenJObject();
      for (RavenJToken item : ar) {
        RavenJToken key = ((RavenJObject) item).get("Key");
        RavenJToken value = ((RavenJObject) item).get("Value");
        o.set(value(String.class, key), value);

      }
      return (U) (Object) o;
    }
    boolean cast = RavenJToken.class.isAssignableFrom(clazz);
    return convert(clazz, token, cast);
  }

  public static <U> Iterable<U> convert(Class<U> clazz, Iterable<RavenJToken> source) {
    boolean cast = RavenJToken.class.isAssignableFrom(clazz);

    List<U> result = new ArrayList<>();
    for (RavenJToken token : source) {
      result.add(convert(clazz, token, cast));
    }
    return result;
  }

  public static <U> U convert(Class<U> clazz, RavenJToken token, boolean cast) {
    if (cast || Object.class.equals(clazz)) {
      // HACK
      return (U) (Object) token;
    }
    if (token == null) {
      return null;
    }

    RavenJValue value;
    try {
      value = (RavenJValue) token;
    } catch (ClassCastException e) {
      throw new ClassCastException("Cannot cast " + token.getType() + " to " + clazz + ".");
    }

    if (value.getValue().getClass().equals(clazz)) {
      return (U) value.getValue();
    }

    //Type targetType = typeof(U);
    //TODO: fixme
    if (!clazz.isPrimitive()) {
      if (value.getValue() == null) {
        try {
          return clazz.newInstance();
        } catch (InstantiationException | IllegalAccessException e) {
          //TODO: implement
        }
      }

      //targetType = Nullable.GetUnderlyingType(targetType);
    }
    if (clazz.equals(UUID.class)) {
      //is it needed?
      if (value.getValue() == null) {
        return (U) UUID.randomUUID();
      }

      return (U) UUID.fromString(value.getValue().toString());
    }
    if (clazz.equals(String.class)) {
      if (value.getValue() == null) {
        try {
          return clazz.newInstance();
        } catch (InstantiationException | IllegalAccessException e) {
          //TODO: implement
        }
      }
      return (U) (Object) value.getValue().toString();
    }
    if (clazz.equals(Date.class) && value.getValue().getClass().equals(String.class)) {
      SimpleDateFormat sdf = new SimpleDateFormat(Constants.RAVEN_S_DATE_FORMAT);
      try {
        return (U) (Object) sdf.parse((String) value.getValue());
      } catch (ParseException e) {
        // TODO implement
      }

    }
    //TODO:
    /*
    if (targetType == typeof(DateTimeOffset) && value.Value is string)
    {
        DateTimeOffset dateTimeOffset;
        if (DateTimeOffset.TryParseExact((string)value.Value, Default.DateTimeFormatsToRead, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out dateTimeOffset))
            return (U)(object)dateTimeOffset;

        return default(U);
    }*/
    //return (U)System.Convert.ChangeType(value.Value, targetType, CultureInfo.InvariantCulture);
    return (U) ConvertUtils.convert(value.getValue(), clazz);
  }

  private static class RavenJTokenIterable<U> implements Iterable<U> {

    private Class<U> clazz;
    private final Iterable<RavenJToken> source;
    private String key;

    public RavenJTokenIterable(Class<U> clazz, Iterable<RavenJToken> source, String key) {
      super();
      this.clazz = clazz;
      this.source = source;
      this.key = key;
    }

    @Override
    public Iterator<U> iterator() {
      return new RavenJTokenIterator<U>();
    }

    private class RavenJTokenIterator<T> implements Iterator<T> {

      Iterator<RavenJToken> sourceIterator = source.iterator();

      @Override
      public boolean hasNext() {
        // TODO Auto-generated method stub
        return sourceIterator.hasNext();
      }

      @Override
      public T next() {
        RavenJToken token = sourceIterator.next();
        if (token instanceof RavenJValue) {
          return (T) convert(clazz, token);
        } else {
          for (U t : token.values(clazz)) {
            return (T) t;
          }
        }

        RavenJObject ravenJObject = (RavenJObject) token;

        RavenJToken value = ravenJObject.get(key);
        if (value != null) {
          return (T) convert(clazz, value);
        }
        return null;
      }

      @Override
      public void remove() {
        throw new IllegalStateException("Deleting elements in iterator is not implemneted!");
      }

    }

  }



}
